service/preinit.go

```diff
@@ -0,0 +1,170 @@
+
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultNodeMdmsFile = "/data/node_mdms.txt"
+)
+
+type PreInitService interface {
+	PreInit() error
+}
+
+func NewPreInitService() PreInitService {
+	return &service{}
+}
+
+type ArrayConfigurationProvider interface {
+	GetArrayConfiguration() ([]*ArrayConnectionData, error)
+}
+
+type FileWriterProvider interface {
+	WriteFile(filename string, data []byte, perm os.FileMode) error
+}
+
+type DefaultArrayConfigurationProvider struct{}
+
+func (s *DefaultArrayConfigurationProvider) GetArrayConfiguration() ([]*ArrayConnectionData, error) {
+	arrayConfig, err := getArrayConfig(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	connectionData := make([]*ArrayConnectionData, 0)
+	for _, v := range arrayConfig {
+		connectionData = append(connectionData, v)
+	}
+
+	return connectionData, nil
+}
+
+type DefaultFileWriterProvider struct{}
+
+func (s *DefaultFileWriterProvider) WriteFile(filename string, data []byte, perm os.FileMode) error {
+	return os.WriteFile(filename, data, perm)
+}
+
+var (
+	arrayConfigurationProviderImpl ArrayConfigurationProvider = &DefaultArrayConfigurationProvider{}
+	fileWriterProviderImpl         FileWriterProvider         = &DefaultFileWriterProvider{}
+	nodeMdmsFile                                              = defaultNodeMdmsFile
+)
+
+func (s *service) PreInit() error {
+	Log.Infof("PreInit running")
+
+	arrayConfig, err := arrayConfigurationProviderImpl.GetArrayConfiguration()
+	if err != nil {
+		return fmt.Errorf("unable to get array configuration: %v", err)
+	}
+
+	labelKey, err := getLabelKey(arrayConfig)
+	if err != nil {
+		return fmt.Errorf("unable to get zone label key: %v", err)
+	}
+
+	var mdmData string
+
+	if labelKey == "" {
+		Log.Debug("No zone key found, will configure all MDMs")
+		sb := strings.Builder{}
+		for _, connectionData := range arrayConfig {
+			if connectionData.Mdm != "" {
+				if sb.Len() > 0 {
+					sb.WriteString("\\&")
+				}
+				sb.WriteString(connectionData.Mdm)
+			}
+		}
+		mdmData = sb.String()
+	} else {
+		Log.Infof("Zone key detected, will configure MDMs for this node, key: %s", labelKey)
+		nodeLabels, err := s.GetNodeLabels(context.Background())
+		if err != nil {
+			return fmt.Errorf("unable to get node labels: %v", err)
+		}
+
+		zone, ok := nodeLabels[labelKey]
+
+		if ok && zone == "" {
+			Log.Errorf("node key found but zone is missing, will not configure MDMs for this node, key: %s", labelKey)
+		}
+
+		if zone != "" {
+			Log.Infof("zone found, will configure MDMs for this node, zone: %s", zone)
+			mdmData, err = getMdmList(arrayConfig, labelKey, zone)
+			if err != nil {
+				return fmt.Errorf("unable to get MDM list: %v", err)
+			}
+		}
+	}
+
+	Log.Infof("Saving MDM list to %s, MDM=%s", nodeMdmsFile, mdmData)
+	err = fileWriterProviderImpl.WriteFile(nodeMdmsFile, []byte(fmt.Sprintf("MDM=%s\n", mdmData)), fs.FileMode(0o444))
+	return err
+}
+
+// Returns a string with the comma separated list of MDM addresses given a
```

Comment:
Should this be updated to say ampersand separated?

Reply:
Yes it should be. Fixed.